### ISSUE TITLE
how to process mutli-input model using DeepExplainer

### ISSUE THREAD

**Issue by bingojojstu** [shap/shap#439]
Hi, 

I have a multi-inputs and multi-outputs model. 

i.e. input: a=(?,5), b=(?,6,7) ; output: 10 digit multi-class target (like MNIST)

so that pred=model.predict([a]+[b]), which works as intended.

However, shap.DeepExplainer(model, [a]+[b]) doesn't work.

Let's say we test on 100 samples, I got the following error
`ValueError: Cannot feed value of shape (100,) for Tensor 'feature1', which has shape '(?, 1)'`

Could you please point out what I did wrong here?

**Comment by slundberg** [comment]
I am surprised you didn't hit an assert that checks for multi-outputs. They are not supported directly right now, you currently have to break up the model into two heads (one for each output) then explain each one.

**Comment by YuRiTan** [comment]
I had the same issue when using multiple inputs and outputs. @bingojojstu your features seem to be in a wrong shape, since you are expecting a `(batch_size, 1)` or `(?, 1)`, but provide a `(batch_size,)`.  With Numpy for example you could do this using `.reshape(-1,1)`. (or `tf.reshape` for Tensorflow, or `view` in PyTorch). With this, I am able to use multiple inputs successfully the same way as you did @bingojojstu.

Multiple output runs but doesn't show all outputs like you've mentioned above. It looks like it's returning the last element of the outputs (list) when using multiple outputs in your (keras) model. For example: the output value that the `force_plot` shows is the last value in the outputs (list). 

In my use case, I have multiple outputs (multiple quantile regression outputs with different quantile loss functions) but I'm only interested in 1 output value for visualization/reasoning purposes using shap. Do you have any idea how to tackle this?

To reproduce, I've created a piece of dummy code (hope this helps):
```python
import shap
import tensorflow as tf
import numpy as np

inp = tf.keras.layers.Input(shape=(2,), name="left_inp")
inp2 = tf.keras.layers.Input(shape=(1,), name="other_inp")
main_inp = tf.keras.layers.Concatenate()([inp, inp2])
intermediate_outp = tf.keras.layers.Dense(10)(main_inp)
outp = [tf.keras.layers.Dense(1)(intermediate_outp) for _ in range(2)] # to simulate multiple outputs
simple_model = tf.keras.Model(
    inputs=[inp, inp2],
    outputs=outp,
)
simple_model.compile(loss='mse', optimizer='adam')
simple_model.summary()

x_train_1 = np.random.randint(0, 100, size=(100,2))
x_train_2 = np.random.randint(0, 100, size=(100,1))
x_test_1 = np.random.randint(0, 100, size=(10,2))
x_test_2 = np.random.randint(0, 100, size=(10,1))

# Skipping the simple_model.fit(...) for now

explainer = shap.DeepExplainer(simple_model, [x_train_1, x_train_2])
shap_values = explainer.shap_values([x_test_1, x_test_2])

shap.initjs()

i = 5
test_inp = [x_test_1[i].reshape(1,-1), x_test_2[i].reshape(1,-1)]
pred = simple_model.predict(test_inp)
print("model outputs: {}".format(pred))
shap.force_plot(explainer.expected_value[0], np.concatenate([shap_values[0][0], shap_values[0][1]], axis=1)[i])
```
which outputs:
```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
left_inp (InputLayer)           (None, 2)            0                                            
__________________________________________________________________________________________________
other_inp (InputLayer)          (None, 1)            0                                            
__________________________________________________________________________________________________
concatenate_3 (Concatenate)     (None, 3)            0           left_inp[0][0]                   
                                                                 other_inp[0][0]                  
__________________________________________________________________________________________________
dense_8 (Dense)                 (None, 10)           40          concatenate_3[0][0]              
__________________________________________________________________________________________________
dense_9 (Dense)                 (None, 1)            11          dense_8[0][0]                    
__________________________________________________________________________________________________
dense_10 (Dense)                (None, 1)            11          dense_8[0][0]                    
==================================================================================================
Total params: 62
Trainable params: 62
Non-trainable params: 0
__________________________________________________________________________________________________

model outputs: [array([[14.550279]], dtype=float32), array([[8.091007]], dtype=float32)]
```
![image](https://user-images.githubusercontent.com/2980353/59336043-586a2580-8ced-11e9-8155-c3884be692ad.png)

Which takes the last output from the multi-output output (because of ![this line](https://github.com/slundberg/shap/blob/9866ae7d22eaa149abfcbd159c4e9a3fe9ed67d4/shap/explainers/deep/deep_tf.py#L85).)

**EDIT 12-06-2019**
FYI, Instead of focusing on the SHAP package, I managed to solve it in a different way by looking at the Keras model itself. You can simply create a new `Model` instance by specifying the inputs and outputs (or single output in this case) from another model:
```python
simple_model_only_first_output = tf.keras.Model(
    inputs=simple_model.inputs,
    outputs=simple_model.outputs[0],  # specifying a single output for shap usage
)

explainer = shap.DeepExplainer(simple_model_only_first_output, [x_train_1, x_train_2])
shap_values = explainer.shap_values([x_test_1, x_test_2])

shap.initjs()

i = 5
test_inp = [x_test_1[i].reshape(1,-1), x_test_2[i].reshape(1,-1)]
pred = simple_model_only_first_output.predict(test_inp)
print("model outputs: {}".format(pred))
shap.force_plot(explainer.expected_value[0], np.concatenate([shap_values[0][0], shap_values[0][1]], axis=1)[i])
```
which outputs:
```
model outputs: [[14.550279]]
```
![image](https://user-images.githubusercontent.com/2980353/59336263-c0b90700-8ced-11e9-8aaf-81f8fd6d6890.png)

**Comment by bingojojstu** [comment]
That's what i about to say, if you follow how intermediate layer output is calculated you should be able to solve your issue.

**Comment by chillburg** [comment]
Hi! Thanks for giving an example, but when I am trying to re-run any of them, I get an error below.
AttributeError: 'list' object has no attribute 'sum'

Maybe someone knows what is the problem? Was very happy to find this thread as it should fix something I am struggling with for a long time.

Version 3.6.8 (default, Oct  7 2019, 12:59:55) [GCC 8.3.0], working in Colab

![image](https://user-images.githubusercontent.com/58375167/69906712-ec2ccc80-13c7-11ea-8e16-4362c1692e49.png)

P.S. If someone is struggling with multi-input models and SHAP, you can solve this problem with a slice() layer. Basically, you concatenate your data into one chunk, and then slice it back inside the model. Problem solved and SHAP works fine! At least that how it worked out for me.

input = Input(shape=(data.shape[1], ))

x1 = Lambda( lambda x: slice(x, (0, 0 ), (-1, 11)))(input)      # deep continuous variables (11)
x2 = Lambda( lambda x: slice(x, (0, 11), (-1, 1)))(input)       # embedding (1)

embed1 = Embedding(input_dim = sub_grade_unique+2, output_dim = 8)(x2)
embed1 = Reshape(target_shape=(8,))(embed1)

concat = keras.layers.concatenate([x1, embed1])

Big thanks to @grofte for suggesting it!

**Comment by YuRiTan** [comment]
@chillburg, I ran the code again, but it runs without any errors with the following configuration:

Python 3.6.8 
shap 0.30.1
numpy 1.16.5
tensorflow 1.12.3
(on Ubuntu 18.04)

It think it therefore might have something to do with the versions...

To really know why you get this, error I would need some time to dive into this. But by quickly looking at the error, it tries to do a chained `.sum()` operation on a python list, which is not possible. You could therefore try to provide a numpy array instead of a list to the `explainer.shap_values()` function, like this:

```
shap_values = explainer.shap_values(np.array([x_test_1, x_test_2]))
```


Hopefully this resolves the error for you. With my setup, both ways work.

**Comment by IdoBl** [comment]
@YuRiTan, When trying to use explainer.shap_values(np.array([x_test_1, x_test_2])) you will get a broadcasting error since x_test_1.shape[1] != x_test_2.shape[1]  
I'm also getting the same problem when additivity is being calculated. in line 284, class TFDeepExplainer:
`diffs = model_output[:, l] - self.expected_value[l] - output_phis[l].sum(axis=tuple(range(1, output_phis[l].ndim)))`
Since output_phis is list, not ndarray.
If I change check_additivity=False the check is ignored and it works fine.
I have multiple input but single output.
Shap version: 0.33.0, but got the same error with Shap 0.30.1
Looks like bug to me. Anyone succeed solving this?
Thanks.

**Comment by slundberg** [comment]
The DeepExplainer is getting updated for TF 2.0 right now, so once that lands we can tackle this.

**Comment by mihailyanchev** [comment]
Hi guys, I am struggling with a multi-input model as well, but run into another issue. Would be happy to hear any opinions about possible solutions. I am using the tf-nightly build 2.4.0-dev20200728 and shap 0.37.0.

I have a model with two inputs (None, None, 12) and (None, 7). I have done the following to fix them and get the shap values:
```
from keras.layers import Input
from keras.models import Model

newInput1 = Input(batch_shape=(44, 16, 12))
newInput2 = Input(batch_shape=(44, 7))
newOutputs = original_model([newInput1, newInput2])
newModel = Model([newInput1, newInput2],newOutputs)

explainer = shap.DeepExplainer(newModel, [x_train1, x_train2])
shap_values = explainer.shap_values([x_test1, x_test2])
```

This returns several weird warnings and the error, which I saw in other threads as well:
```
WARNING:tensorflow:Model was constructed with shape (44, 16, 12) for input Tensor("input_1:0", shape=(44, 16, 12), dtype=float32), but it was called on an input with incompatible shape (88, 16, 12).
WARNING:tensorflow:Model was constructed with shape (44, 7) for input Tensor("input_2:0", shape=(44, 7), dtype=float32), but it was called on an input with incompatible shape (88, 7).

AttributeError: 'TFDeep' object has no attribute 'between_tensors'
```

I see what the _init_between_tensors is doing and I assume there is something unusual about the graph (I am attaching that as well). 
![2nd_step_model](https://user-images.githubusercontent.com/34907806/99145767-0cf23e80-267a-11eb-86e9-a4835149e6c4.png)
Any feedback will be highly appreciated!

**Comment by Omid149** [comment]
Hi @mihailyanchev, any update?

**Comment by Mushahid2521** [comment]
> 

Hello,
It helps. I was trying to do the beeswarm plot but can't do it passing, shap_values[class_idx][inputX] into it. How to do it?

**Comment by tahsin314** [comment]
> mihailyanchev

I'm also struggling with the similar issue. Any update?
